lib/scopes/scopes.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	if scope == exclusiveChildGlobSuffix {
+		// this is just a matcher for any child of root
+		return nil
+	}
+
+	return StrongValidate(strings.TrimSuffix(scope, exclusiveChildGlobSuffix))
+}
+
+// DescendingSegments produces an iterator over the segments of a scope in descending order.
+// e.g. `DescendingSegments("/a/b/c")` will result in an iterator that returns `a`, `b`, and
+// `c` in that order. `DescendingSegments("/")` will return an empty iterator.
+//
+// Note that this function does not perform validation and is deliberately more relaxed about
+// its inputs than our validation functions allow.
+func DescendingSegments(scope string) iter.Seq[string] {
+	if scope == "" || scope == separator {
+		return func(yield func(string) bool) {}
+	}
+	return strings.SplitSeq(strings.TrimSuffix(strings.TrimPrefix(scope, "/"), "/"), "/")
+}
+
+// Join joins the given segments into a single scope string. Note that this function
+// does not perform validation and will produce invalid scopes if one or more segments
+// are invalid.
+func Join(segments ...string) string {
+	scope := separator + strings.Join(segments, separator)
+
+	// a tricky bit of how scope splitting/joining works is that an empty scope component
+	// as the last component is represented by `/aa//` not `/aa/`, because we chose not to assign trailing
+	// separators any meaning (i.e. `/aa/` does not imply the existence of an empty scope segment
+	// after `aa`).  Trailing separators and empty scope segments are both considered invalid, but adding
+	// this behavior ensures that our splitting/joining logic cannot cannot change the meaning of a scope
+	// value when applied to one-another's outputs. If we don't add this logic, then
+	// DescendingSegments(Join([]string{"aa", ""}...)) will produce []string{"aa"}. We could just as easily amend
+	// `DescendingSegments` to emit an empty segment in the event of a trailing separator, but
+	// we judged that to be the higher risk behavior since that would effectively "move" a value to a new
+	// scope in the event of a trailing separator making it past validation. Our chosen behavior only "moves" an assignment
+	// in the event of a double-separator getting through, which is a much more visually obvious and intuitively incorrect
+	// value, and therefore less likely to be hit during processing of user input.
+	if len(segments) > 0 && segments[len(segments)-1] == "" {
+		scope += separator
+	}
+
+	return scope
+}
+
+// Relationship describes the relationship between two scopes, as determined by the [Compare] function. Note
+// that direct use of this type in access-control logic is discouraged, as it is easier to accidentally
+// misuse than the provided helpers (e.g. [PolicyScope]).
+type Relationship int
+
+const (
+	// Orhogonal indicates that the scopes are not related.
+	Orthogonal Relationship = iota
+	// Equivalent indicates that the scopes are equal.
+	Equivalent
+	// Ancestor indicates that one scope is an ancestor of another.
```

Comment:
I think it'd be helpful to explicitly state within this GoDoc that "Ancestor"/"Descendant" will also include "grandparents"/"grandchildren" and that it doesn't only represent a direct parent/child.